examples/MidiKeyboard/KeyboardWidget.hpp

```diff
@@ -0,0 +1,486 @@
+    bool fPressed;
+    Image fImageNormal;
+    Image fImageDown;
+    int fIndex;
+};
+
+class KeyboardWidget : public Widget
+{
+public:
+    class Callback
+    {
+    public:
+        virtual ~Callback() {}
+        virtual void keyboardKeyPressed(const uint keyIndex) = 0;
+        virtual void keyboardKeyReleased(const uint keyIndex) = 0;
+    };
+
+    KeyboardWidget(Window& parent)
+        : Widget(parent),
+          fCallback(nullptr),
+          fMouseDown(false),
+          fHeldKey(nullptr)
+    {
+        fSVGs[kWhiteKeyResourceIndex].loadFromMemory(MidiKeyboardResources::white_keyData,
+                                                     MidiKeyboardResources::white_keyDataSize,
+                                                     1.0f);
+
+        fSVGs[kWhiteKeyPressedResourceIndex].loadFromMemory(MidiKeyboardResources::white_key_pressedData,
+                                                            MidiKeyboardResources::white_key_pressedDataSize,
+                                                            1.0f);
+
+        fSVGs[kBlackKeyResourceIndex].loadFromMemory(MidiKeyboardResources::black_keyData,
+                                                     MidiKeyboardResources::black_keyDataSize,
+                                                     1.0f);
+
+        fSVGs[kBlackKeyPressedResourceIndex].loadFromMemory(MidiKeyboardResources::black_key_pressedData,
+                                                    MidiKeyboardResources::black_key_pressedDataSize,
+                                                    1.0f);
+
+        for (int i = 0; i < kResourcesCount; ++i)
+        {
+            fImages[i].loadFromSVG(fSVGs[i]);
+        }
+
+        const int whiteKeysTotalSpacing = kWhiteKeySpacing * kWhiteKeysCount;
+
+        const uint width = fImages[kWhiteKeyResourceIndex].getWidth() * kWhiteKeysCount + whiteKeysTotalSpacing;
+        const uint height = fImages[kWhiteKeyResourceIndex].getHeight();
+
+        setSize(width, height);
+
+        setupKeyLookupTable();
+        setKeyImages();
+        positionKeys();
+    }
+
+    /**
+       Set the 'pressed' state of a key in the keyboard.
+     */
+    void setKeyPressed(const uint keyIndex, const bool pressed, const bool sendCallback = false)
+    {
+        DISTRHO_SAFE_ASSERT_RETURN(keyIndex < kKeyCount, )
+
+        PianoKey& key = *fKeysLookup[keyIndex];
+
+        if (key.isPressed() == pressed)
+            return;
+
+        key.setPressed(pressed);
+
+        if (fCallback != nullptr && sendCallback)
+        {
+            if (pressed)
+            {
+                fCallback->keyboardKeyPressed(keyIndex);
+            }
+            else
+            {
+                fCallback->keyboardKeyReleased(keyIndex);
+            }
+        }
+
+        repaint();
+    }
+
+    void setCallback(Callback* callback)
+    {
+        fCallback = callback;
+    }
+
+    /**
+       Draw the piano keys.
+     */
+    void onDisplay() override
```

Comment:
Right now, the white and black keys are handled separately, to simplify z-order handling.

If these were widgets, I think it would've been possible to just instantiate the white keys first, and the black ones last. z-order would've been correct without any further adjustments.